src/System.Text.Primitives/System/Text/Encoders/UnicodeScalar.cs

```diff
@@ -0,0 +1,246 @@
+                utf16[0] = (char)(0xD7C0U + ((uint)Value >> 10));
+                utf16[1] = (char)(0xDC00U | ((uint)Value & 0x3FFU));
+            }
+            else
+            {
+                throw new ArgumentException(
+                    message: "Argument is not long enough to hold output value.",
+                    paramName: nameof(utf16));
+            }
+        }
+
+        /// <summary>
+        /// Copies the UTF-8 code unit representation of this scalar to an output buffer.
+        /// The buffer must be large enough to hold the required number of <see cref="byte"/>s.
+        /// The <see cref="Utf8CodeUnitCount"/> property gives the required output buffer length.
+        /// </summary>
+        public void CopyUtf8CodeUnitsTo(Span<byte> utf8)
+        {
+            // See the Unicode Standard, Table 3-6.
+            Debug.Assert(IsValid);
+
+            if ((uint)Value < 0x80U && utf8.Length >= 1)
+            {
+                // Single UTF-8 code unit
+                utf8[0] = (byte)Value;
+            }
+            else if ((uint)Value < 0x800U && utf8.Length >= 2)
+            {
+                // Two UTF-8 code units
+                utf8[0] = (byte)(0xC0U | ((uint)Value >> 6));
+                utf8[1] = (byte)(0x80U | ((uint)Value & 0x3FU));
+            }
+            else if ((uint)Value < 0x10000U && utf8.Length >= 3)
+            {
+                // Three UTF-8 code units
+                utf8[0] = (byte)(0xE0U | ((uint)Value >> 12));
+                utf8[1] = (byte)(0x80U | (((uint)Value >> 6) & 0x3FU));
+                utf8[2] = (byte)(0x80U | ((uint)Value & 0x3FU));
+            }
+            else if (utf8.Length >= 4)
+            {
+                // Four UTF-8 code units
+                utf8[0] = (byte)(0xF0U | ((uint)Value >> 18));
+                utf8[1] = (byte)(0x80U | (((uint)Value >> 12) & 0x3FU));
+                utf8[2] = (byte)(0x80U | (((uint)Value >> 6) & 0x3FU));
+                utf8[3] = (byte)(0x80U | ((uint)Value & 0x3FU));
+            }
+            else
+            {
+                throw new ArgumentException(
+                  message: "Argument is not long enough to hold output value.",
+                  paramName: nameof(utf8));
+            }
+        }
+
+        internal static UnicodeScalar CreateWithoutValidation(uint value) => new UnicodeScalar(value);
+
+        public override bool Equals(object other) => ((other is UnicodeScalar) && this.Equals((UnicodeScalar)other));
+
+        public bool Equals(UnicodeScalar other) => (this.Value == other.Value);
+
+        public override int GetHashCode() => Value;
+
+        private static bool IsValidScalar(uint value) =>
+            (value < 0xD800U) || Utf8Util.IsInRangeInclusive(value, 0xE000U, 0x10FFFFU);
+
+        public override string ToString()
+        {
+            // This can be made much more efficient.
+            return "U+" + ((uint)Value).ToString("X4", CultureInfo.InvariantCulture);
```

Comment:
It might be good to return both the numeric and textual representation of the scalar.

Reply:
Are there any immediate concerns with this method being inefficient? I didn’t really bother improving this because I thought it would be used more for debugging than for anything else.

Reply:
Yeah, it's not about efficiency. It's about debug-ability.